crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -446,21 +455,183 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_val: Value =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        let ticket_type = context
+            .session
+            .load_type(&TypeTag::Struct(Box::new(UpgradeTicket::type_())))
+            .map_err(|e| context.convert_vm_error(e))?;
+        check_param_type::<_, _, Mode>(context, 0, &ticket_val, &ticket_type)?;
+        ticket_val.write_bcs_bytes(&mut ticket_bytes);
+        bcs::from_bytes(&ticket_bytes).map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::CommandArgumentError {
+                arg_idx: 0,
+                kind: CommandArgumentError::InvalidBCSBytes,
+            })
+        })?
+    };
+
+    // Make sure the passed-in package ID matches the package ID in the `upgrade_ticket`.
+    if current_package_id != upgrade_ticket.package.bytes {
+        return Err(ExecutionError::from_kind(
+            ExecutionErrorKind::PackageUpgradeError {
+                upgrade_error: PackageUpgradeError::PackageIDDoesNotMatch {
+                    package_id: current_package_id,
+                    ticket_id: upgrade_ticket.package.bytes,
+                },
+            },
+        ));
+    }
+
+    // Check digest.
+    let computed_digest =
+        MovePackage::compute_digest_for_modules_and_deps(&module_bytes, &dep_ids).to_vec();
+    if computed_digest != upgrade_ticket.digest {
+        return Err(ExecutionError::from_kind(
+            ExecutionErrorKind::PackageUpgradeError {
+                upgrade_error: PackageUpgradeError::DigestDoesNotMatch {
+                    digest: computed_digest,
+                },
+            },
+        ));
+    }
+
+    // Check that this package ID points to a package and get the package we're upgrading.
+    let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    // NB: this will substitute in the original package id for the `self` address in all of these modules.
+    let upgraded_package_modules = publish_and_verify_upgraded_modules::<_, _, Mode>(
+        context,
+        &module_bytes,
+        current_package.original_package_id(),
+    )?;
+
+    // Full backwards compatibility except that we allow friend function signatures to change.
+    let check_struct_and_pub_function_linking = true;
+    let check_struct_layout = true;
+    let check_friend_linking = false;
+    check_compatibility(
+        &current_package,
+        &upgraded_package_modules,
+        upgrade_ticket.policy,
+        check_struct_and_pub_function_linking,
+        check_struct_layout,
+        check_friend_linking,
+    )?;
+
+    // Deduplicate any dependency ids.
+    let dep_ids: BTreeSet<_> = dep_ids.into_iter().collect();
+
+    // Read the package dependencies.
+    let dependency_packages = dep_ids
```

Comment:
Making a note that we should use your shiny new `fetch_packages` function here (+ I also left a note in #9096 that we should be deduplicating `dep_ids` on entry into that function as well).

Reply:
Good catch! Indeed I do probably want to use that here...